frame/staking/src/lib.rs

```diff
@@ -939,6 +961,217 @@ where
+	/// be made public.
+	fn is_rewards_claimed(era: EraIndex, validator: &T::AccountId, page: PageIndex) -> bool {
+		ClaimedRewards::<T>::get(era, validator).contains(&page)
+	}
+
+	/// Get exposure info for a validator at a given era and page.
+	///
+	/// This builds a paged exposure from `ExposureOverview` and `ExposurePage` of the
+	/// validator. For older non-paged exposure, it returns the clipped exposure directly.
+	pub(crate) fn get_validator_exposure(
+		era: EraIndex,
+		validator: &T::AccountId,
+		page: PageIndex,
+	) -> Option<ExposureExt<T::AccountId, BalanceOf<T>>> {
+		let overview = <ErasStakersOverview<T>>::get(&era, validator);
+
+		// return clipped exposure if page zero and paged exposure does not exist
+		// exists for backward compatibility and can be removed as part of #13034
+		if overview.is_none() && page == 0 {
+			return Some(ExposureExt::from_clipped(<ErasStakersClipped<T>>::get(era, validator)))
+		}
+
+		// no exposure for this validator
+		if overview.is_none() {
+			return None
+		}
+
+		let overview = overview.unwrap();
+
+		// validator stake is added only in page zero
+		let validator_stake = if page == 0 { overview.own } else { Zero::zero() };
+
+		// since overview is present, paged exposure will always be present except when a
+		// validator has only own stake and no nominator stake.
+		let exposure_page =
+			<ErasStakersPaged<T>>::get((era, validator, page)).unwrap_or_default();
+
+		// build the exposure
+		Some(ExposureExt {
+			exposure_overview: ExposureOverview { own: validator_stake, ..overview },
+			exposure_page,
+		})
+	}
+
+	/// Get complete exposure of the validator at a given era.
+	pub(crate) fn get_non_paged_validator_exposure(
+		era: EraIndex,
+		validator: &T::AccountId,
+	) -> Exposure<T::AccountId, BalanceOf<T>> {
+		let overview = <ErasStakersOverview<T>>::get(&era, validator);
+
+		if overview.is_none() {
+			return ErasStakers::<T>::get(era, validator)
+		}
+
+		let overview = overview.unwrap();
+
+		if overview.page_count == 0 {
+			return Exposure { total: overview.total, own: overview.own, others: vec![] }
+		}
+
+		let mut others = Vec::with_capacity(overview.nominator_count as usize);
+		for page in 0..overview.page_count {
+			let nominators = <ErasStakersPaged<T>>::get((era, validator, page));
+			others.append(&mut nominators.map(|n| n.others).defensive_unwrap_or_default());
+		}
+
+		Exposure { total: overview.total, own: overview.own, others }
+	}
+
+	/// Returns the number of pages of exposure a validator has for the given era.
+	///
+	/// This will always return at minimum one count of exposure to be backward compatible to
+	/// non-paged reward payouts.
+	pub(crate) fn get_page_count(era: EraIndex, validator: &T::AccountId) -> PageIndex {
+		<ErasStakersOverview<T>>::get(&era, validator)
+			.map(|overview| {
+				if overview.page_count == 0 && overview.own > Zero::zero() {
+					// this means the validator has their own stake but no nominators
```

Comment:
quite confusing why if a validator only has their self-stake, this function returns 1. Why we ended up coding it like this? I suspect the call site might have unreasonable expectations that leaked here.

Reply:
Good point. I agree its not clean and will fix it.

The `call` to payout stakers pays the validator stake only when the page is 1. But as you pointed out it should be handled by the calling function to take care of paying out the validator stake in all these cases instead of an underlying function lying about the page count.

Reply:
On further digging, may be this is not too bad. With this PR, the expectation is that the caller has to specify the page to claim rewards for. If a validator has no nominators (therefore no nominator pages), but their own stake is exposed, that still needs to be paid in a page, hence one page of exposure.

I think the fact that we differentiate between own and other stake of the validator adds to this confusion. I have added a little bit more explanation in the comment for this.